In [6]:
import pandas as pd

from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LinearRegression

from sklearn.metrics import accuracy_score, confusion_matrix

In [4]:
from google.colab import files
uploaded = files.upload()

Saving Dataset 2.csv to Dataset 2.csv


In [5]:
df = pd.read_csv("Dataset 2.csv")
df.head()

,UserID,Age,Gender,SubscriptionType,WatchHoursPerWeek,DevicesUsed,FavoriteGenre,AdClicks,MonthlySpend,SubscriptionRenewed
0,1001,22,Female,Basic,23,5,Comedy,13,353,No
1,1002,55,Male,Basic,9,4,Drama,14,317,Yes
2,1003,49,Male,Basic,8,3,Comedy,16,309,No
3,1004,39,Female,Premium,19,5,Drama,45,833,Yes
4,1005,38,Female,Premium,23,5,Sci-Fi,24,804,Yes


In [7]:
print(df.shape)

(750, 10)


In [8]:
print(df.columns)

Index(['UserID', 'Age', 'Gender', 'SubscriptionType', 'WatchHoursPerWeek',
       'DevicesUsed', 'FavoriteGenre', 'AdClicks', 'MonthlySpend',
       'SubscriptionRenewed'],
      dtype='object')


Numerical :
UserID
Age
WatchHoursPerWeek
DevicesUsed
AdClicks
MonthlySpend

Categorical:
Gender
SubscriptionType
FavoriteGenre
SubscriptionRenewed

In [9]:
print(df.isnull().sum())

UserID                 0
Age                    0
Gender                 0
SubscriptionType       0
WatchHoursPerWeek      0
DevicesUsed            0
FavoriteGenre          0
AdClicks               0
MonthlySpend           0
SubscriptionRenewed    0
dtype: int64


In [10]:
print(df["Age"].mean())

41.824


In [11]:
print(df["WatchHoursPerWeek"].mean())

14.236


In [12]:
print(df["MonthlySpend"].mean())

689.9053333333334


In [13]:
print(df["SubscriptionType"].value_counts())

SubscriptionType
Basic      342
Premium    279
VIP        129
Name: count, dtype: int64


In [14]:
renew_pct = (df["SubscriptionRenewed"]=="Yes").mean()*100
print(renew_pct)

46.266666666666666


In [15]:
df_encoded = df.copy()

le = LabelEncoder()

for col in ["Gender",
            "SubscriptionType",
            "FavoriteGenre",
            "SubscriptionRenewed"]:
    df_encoded[col] = le.fit_transform(df_encoded[col])

df_encoded.head()

,UserID,Age,Gender,SubscriptionType,WatchHoursPerWeek,DevicesUsed,FavoriteGenre,AdClicks,MonthlySpend,SubscriptionRenewed
0,1001,22,0,0,23,5,1,13,353,0
1,1002,55,1,0,9,4,2,14,317,1
2,1003,49,1,0,8,3,1,16,309,0
3,1004,39,0,1,19,5,2,45,833,1
4,1005,38,0,1,23,5,5,24,804,1


In [16]:
X = df_encoded.drop("SubscriptionRenewed", axis=1)
y = df_encoded["SubscriptionRenewed"]

In [17]:
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.20,
    random_state=42
)

In [18]:
dt = DecisionTreeClassifier(random_state=42)

dt.fit(X_train, y_train)

DecisionTreeClassifier(random_state=42)

In [19]:
y_pred_dt = dt.predict(X_test)

dt_accuracy = accuracy_score(y_test, y_pred_dt)

print(dt_accuracy)

0.5666666666666667


In [20]:
cm = confusion_matrix(y_test, y_pred_dt)

print(cm)

[[50 32]
 [33 35]]


In [21]:
knn = KNeighborsClassifier(n_neighbors=5)

knn.fit(X_train, y_train)

y_pred_knn = knn.predict(X_test)

In [22]:
knn_accuracy = accuracy_score(y_test, y_pred_knn)

print(knn_accuracy)

0.6


In [23]:
X_reg = df_encoded.drop("MonthlySpend", axis=1)
y_reg = df_encoded["MonthlySpend"]

X_train_reg, X_test_reg, y_train_reg, y_test_reg = train_test_split(
    X_reg,
    y_reg,
    test_size=0.20,
    random_state=42
)

lr = LinearRegression()

lr.fit(X_train_reg, y_train_reg)

LinearRegression()

In [24]:
new_user = X_reg.iloc[[0]]

prediction = lr.predict(new_user)

print(prediction)

[391.38516281]


Business Reflection Questions

1.
Subscription Type
 Monthly Spending
 Watch Hours Per Week
 Devices Used
 Advertisement Clicks

2.
Because the target variable has two categories:
Yes
No
The model predicts a class rather than a continuous value.

3.
Because spending is a continuous numeric value:

₹353
₹833
₹517

The model predicts a number rather than a category.

4.
KNN performed better.

5.
Identify users likely to cancel subscriptions.
Offer personalized discounts.
Recommend content based on viewing habits.
Improve retention campaigns.
Increase revenue through targeted marketing.